section about atlas setFeature function (#4222)
qgis/QGIS-Documentation: source/docs/user_manual/print_composer/create_output.rst

.. _Multiple_format_map_series_using_QGIS_2.6: https://sigsemgrilhetas.wordpress.com/2014/11/09/series-de-mapas-com-formatos-multiplos-em-qgis-2-6-parte-1-multiple-format-map-series-using-qgis-2-6-part-1

.. _relations_in_atlas:

Use project defined relations for atlas creation

> -----------------------------------------------

Reviewer: This is what is making travis complain.

```suggestion
------------------------------------------------
```

For users with decent HTML and Javascript knowledge there is the possibility to

> operate on GeoJSON objects and so use project defined relations from the QGIS project.

Reviewer 2: ```suggestion
operate on GeoJSON objects and use project defined relations from the QGIS project.
```

> The big difference between this approach versus just using expressions

Reviewer 2: ```suggestion
The difference between this approach and using expressions
```

directly inserted into the HTML is that it gives you a full,
unstructured GeoJSON feature to work with. This means that you can use

> existing javascript libraries and functions which operate on geojson

Reviewer 2: ```suggestion
existing Javascript libraries and functions, which operate on GeoJSON
```

feature representations.
The following code includes all related child features from the defined relation.
Using the JavaScript ``setFeature`` function it allows you to make flexible HTML
which represents relations in whatever format you like (lists, tables, etc).
In the code sample we create a dynamic bullet list of related child features.

.. code::

   <script>
   <div id="name"></div>
   <div id="my_children"></div>

   <script type="text/javascript">

>       function setFeature(feature)

Reviewer 2: This is where you handle the feature (the feature is the argument of the function). So you need to have the feature before you call this function.

      {
      document.getElementById('name').innerHTML = feature.properties.NAME;
      document.getElementById('my_children').innerHTML = ''; 

>       //clear existing relation contents

Reviewer 2: @SrNetoChan, 'name' and 'my_children' are the ids of the two HTML <div> elements defined above. The result of the above code is to set the first <div>s text to the name attribute of the feature, and the second <div>s text to blank (empty).

      feature.properties.my_relation.forEach(function(child_feature) {
      // for each related child feature, create a LI element with the feature's name

>                                                                      var node = document.createElement("li");

Reviewer 2: ```suggestion
        var node = document.createElement("li");
```

>                                                                      node.appendChild(document.createTextNode(child_feature.NAME));

Reviewer 2: ```suggestion
        node.appendChild(document.createTextNode(child_feature.NAME));
```

>                                                                      document.getElementById('my_children').appendChild(node);

Reviewer 2: ```suggestion
        document.getElementById('my_children').appendChild(node);
```

                                                                     });
      }
   </script>

> During the atlas creation described in the previous section there will be an iteration over

Reviewer 2: ```suggestion
During the atlas creation there will be an iteration over
```

> the coverage layer containing the parent feature. On every page you will see a bullet list of

Reviewer: ```suggestion
the coverage layer containing the parent features. On every page, you will see a bullet list of
```

> the coverage layer containing the parent feature. On every page you will see a bullet list of

Reviewer 2: ```suggestion
the layer containing the parent feature. On every page you will see a bullet list of
```

Reviewer: I really like the precision of mentioning that the parent feature comes from the current atlas coverage layer. Otherwise it would not work.

.. _Multiple_format_map_series_using_QGIS_2.6: https://sigsemgrilhetas.wordpress.com/2014/11/09/series-de-mapas-com-formatos-multiplos-em-qgis-2-6-parte-1-multiple-format-map-series-using-qgis-2-6-part-1

.. _relations_in_atlas:

Use project defined relations for atlas creation
------------------------------------------------

For users with HTML and Javascript knowledge it's possible to operate on
GeoJSON objects and use project defined relations from the QGIS project.
The difference between this approach and using expressions
directly inserted into the HTML is that it gives you a full,
unstructured GeoJSON feature to work with. This means that you can use the
existing Javascript libraries and functions which operate on GeoJSON
feature representations.
The following code includes all related child features from the defined relation.
Using the JavaScript ``setFeature`` function it allows you to make flexible HTML
which represents relations in whatever format you like (lists, tables, etc).
In the code sample, we create a dynamic bullet list of the related child features.

.. code::

   <script>
   <div id="name"></div>
   <div id="my_children"></div>

>    </script>

Reviewer 3: I inserted this line, since script was not closed but I don't have a decent HTML knowledge

Reviewer 2: The <div> tags should not be part of the <script> tag.  So the <script> tag should either be removed or "filled" with content.

Reviewer 3: Fixed. I removed the <script> tags, easier for me than filling it with anything ;)
I also tried to add comments and avoid using the same field for the layers. Would appreciate your feedback. Thanks

.. _Multiple_format_map_series_using_QGIS_2.6: https://sigsemgrilhetas.wordpress.com/2014/11/09/series-de-mapas-com-formatos-multiplos-em-qgis-2-6-parte-1-multiple-format-map-series-using-qgis-2-6-part-1

.. _relations_in_atlas:

Use project defined relations for atlas creation
------------------------------------------------

For users with HTML and Javascript knowledge it's possible to operate on
GeoJSON objects and use project defined relations from the QGIS project.
The difference between this approach and using expressions
directly inserted into the HTML is that it gives you a full,
unstructured GeoJSON feature to work with. This means that you can use the
existing Javascript libraries and functions which operate on GeoJSON
feature representations.

The following code includes all related child features from the defined relation.
Using the JavaScript ``setFeature`` function it allows you to make flexible HTML
which represents relations in whatever format you like (lists, tables, etc).
In the code sample, we create a dynamic bullet list of the related child features.

.. code:: html

>    // Declare the two layers we will use

Reviewer 2: ```suggestion
   // Declare the two HTML div elements we will use for the parent feature id and
   // information about the children
```

   <div id="parent"></div>
   <div id="my_children"></div>

   <script type="text/javascript">
      function setFeature(feature)
      {

>       // Show the parent feature's identifier (using field "ID")

Reviewer 3: ```suggestion
      // Show the parent feature's identifier (using "ID" field)
```

      document.getElementById('parent').innerHTML = feature.properties.ID;

>       document.getElementById('my_children').innerHTML = ''; 

Reviewer 2: ```suggestion
      //clear existing relation contents
      document.getElementById('my_children').innerHTML = ''; 
```

>       //clear existing relation contents

Reviewer 2: ```suggestion
```

      feature.properties.my_relation.forEach(function(child_feature) {
      // for each related child feature, create a list element

>       // with the feature's name (using field "NAME")

Reviewer 2: ```suggestion
      // with the feature's name (using "NAME" field)
```
